autoprecompiles/src/lib.rs

```diff
@@ -1048,24 +1052,287 @@ fn try_compute_opcode_map<T: FieldElement>(
         .collect())
 }
 
+fn powdr_solver_optimize<P: FieldElement>(
+    mut symbolic_machine: SymbolicMachine<P>,
+) -> SymbolicMachine<P> {
+    let constraint_system = ConstraintSystem {
+        algebraic_constraints: symbolic_machine
+            .constraints
+            .iter()
+            .map(|constraint| algebraic_to_quadratic_symbolic_expression(&constraint.expr))
+            .collect(),
+        bus_interactions: symbolic_machine
+            .bus_interactions
+            .iter()
+            .map(|bus_interaction| BusInteraction {
+                bus_id: SymbolicExpression::Concrete(P::from(bus_interaction.id)).into(),
+                payload: bus_interaction
+                    .args
+                    .iter()
+                    .map(|arg| algebraic_to_quadratic_symbolic_expression(arg))
+                    .collect(),
+                multiplicity: algebraic_to_quadratic_symbolic_expression(&bus_interaction.mult),
+            })
+            .collect(),
+    };
+    // Largely copied from https://github.com/powdr-labs/powdr/pull/2643
+    match Solver::new(constraint_system).solve() {
+        Err(_) => {
+            log::error!("Error while QSE-optimizing. This is usually the case when the constraints are inconsistent.");
+        }
+        Ok(SolveResult {
+            simplified_constraint_system,
+            ..
+        }) => {
+            symbolic_machine.constraints = symbolic_machine
+                .constraints
+                .iter()
+                .zip_eq(simplified_constraint_system.algebraic_constraints)
+                .filter_map(|(identity, simplified)| {
+                    println!("Before: {identity}");
+
+                    let expr = quadratic_symbolic_expression_to_algebraic(&simplified);
+                    println!("After: {expr}");
+
+                    if expr == P::zero().into() {
+                        println!("  => Skipping\n");
```

Comment:
In my pilopt PR I noticed that there, we cannot just remove identities because some columns from the identities are referenced from outside the machine. I would assume this is the same here.

Reply:
How? The machine only communicates via bus interactions, no? I think in your PR you didn't pass bus interactions (and lookups, permutations, connect constraints, ...) to the solver, but in this case the solver has access to all constraints within the machine, and another machine would only be able to access them via the bus interactions.